### ISSUE TITLE
[BUG] Parallel orchestration not working for multiple devices

### ISSUE THREAD

**Issue by stephannieu** [saltstack/salt#61208]
**Description**
We have configured an orchestrator file to run multiple commands for multiple switches. We're trying to run 17 commands on more than 1 device but the orchestrator is running them in just the first one specified.

**Setup**
Master
-------
/srv/salt/nxos_commands

```
{% set target = salt['pillar.get']('target') %}
{% for cmd in ['show version','show interfaces description','show ip bgp all','show module','show vlan','show mac address','show ip arp','show interfaces status','show cdp nei','show etherchannel summary','show file systems','show bootvar','show spanning-tree','show inventory','show redundancy','show switch virtual','show run'] %}
run {{ cmd }}:
  salt.function:
    - name: net.cli
    - tgt: '{{ target }}'
    - tgt_type: "compound"
    - arg:
      - {{ cmd }}
    - parallel: True
{% endfor %}
```

Pillar config is stored in an external DB.

- [ ] on-prem machine
- [X] VM (Virtualbox, KVM, etc. please specify)
- [ ] VM running on a cloud service, please be explicit and add details
- [ ] container (Kubernetes, Docker, containerd, etc. please specify)
- [ ] or a combination, please be explicit
- [ ] jails if it is FreeBSD

  
**Steps to Reproduce the behavior**
salt-run state.orch nxos_ssh_commands_org pillar=’{’target’: ‘L@nxs2-ork3,nxs1-ork1’}'
salt-run state.orch nxos_ssh_commands_org pillar=’{’target’: ‘nxs2-ork3,nxs1-ork1’}'

Both commands have same result (run all commands on just the first device on the list)

**Expected behavior**
Commands should run for all specified devices in parallel.

**Screenshots**
N/A

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```
---MASTER---
Salt Version:
          Salt: 3003.3
 
Dependency Versions:
          cffi: 1.14.6
      cherrypy: unknown
      dateutil: Not Installed
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.0.1
       libgit2: Not Installed
      M2Crypto: 0.35.2
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.20
      pycrypto: Not Installed
  pycryptodome: 3.10.1
        pygit2: Not Installed
        Python: 3.6.8 (default, Nov 16 2020, 16:55:22)
  python-gnupg: Not Installed
        PyYAML: 5.4.1
         PyZMQ: 17.0.0
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.1.4
 
System Versions:
          dist: centos 7 Core
        locale: UTF-8
       machine: x86_64
       release: 3.10.0-1160.42.2.el7.x86_64
        system: Linux
       version: CentOS Linux 7 Core

---MINION---
Salt Version:
          Salt: 3003.1
 
Dependency Versions:
          cffi: 1.14.5
      cherrypy: Not Installed
      dateutil: 2.8.2
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.0.1
       libgit2: Not Installed
      M2Crypto: 0.35.2
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.20
      pycrypto: Not Installed
  pycryptodome: Not Installed
        pygit2: Not Installed
        Python: 3.6.8 (default, Apr  2 2020, 13:34:55)
  python-gnupg: Not Installed
        PyYAML: 5.4.1
         PyZMQ: 17.0.0
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.1.4
 
System Versions:
          dist: centos 7 Core
        locale: UTF-8
       machine: x86_64
       release: 3.10.0-1127.19.1.el7.x86_64
        system: Linux
       version: CentOS Linux 7 Core
```
</details>

**Additional context**
N/A

**Comment by waynew** [comment]
Still need to track down exactly *why* we're seeing this bug but it's definitely happening.

**Comment by waynew** [comment]
This *might* be related to proxy minions. I'm going through this right now with regular minions and I'm not able to reproduce. I'm seeing the same general summary but I'm actually seeing the correct response.

I'm not sure if `net.cli` would support this command, but this is what I'm doing:

```
my orchestration:
  salt.function:
    - name: cmd.run
    - arg:
      - echo "hello from $(cat /etc/hostname)"
    - tgt: '*'
    - parallel: True
```

I'm seeing this output:

```
----------
          ID: my orchestration
    Function: salt.function
        Name: cmd.run
      Result: True
     Comment: Function ran successfully. Function cmd.run ran on 15f0c4ca2fd4, aa3d753f1b3e.
     Started: 20:58:12.460799
    Duration: 230.403 ms
     Changes:
              15f0c4ca2fd4:
                  hello from 15f0c4ca2fd4
              aa3d753f1b3e:
                  hello from aa3d753f1b3e

Summary for 3197a5b65265_master
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time: 230.403 ms

```

If you do the same with `net.cli` do you get a similar sort of output? (Might need to run a different command - looks like [`net.ping`](https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.napalm_network.html#salt.modules.napalm_network.ping) should work)

**Comment by waynew** [comment]
I'm pretty confident at this point that this is not a bug in orchestration at all. I think that what we're seeing here is a failure for the proxy minion to actually respond.

I built a very silly proxy module and put it in `/srv/salt/_proxy/foo.py`:

```
import os

__proxyenabled__ = ["foo"]
def __virtual__():
    return 'foo'

def init(opts):
    ...

def initialized():
    return True

def shutdown(): pass

def ping():
    return alive({})

def alive(opts):
    return not os.path.exists('/tmp/dead')

def grains():
    return {'some': 'cool grains'}

def shutdown():
    ...
def cool(input):
    return str(input)
```

and built the following equally silly module and dumped it in `/srv/salt/_modules/cool.py`

```
import os
import time

def beans(sleep=0):
  if os.path.exists('/tmp/dead'):
    # exit('bye!')  # this will make it fail
    pass
  time.sleep(sleep)
  return sleep
```

I ran `salt myproxy\* saltutil.sync_modules` and `salt myproxy\* saltutil.sync_proxymodules` to put them on my proxy minions, and then used an orchestration similar to yours/previous one I used:

```
{% for i in range(5) %}
my orchestrate test {{ i }}:
  salt.function:
    - name: cool.beans:
    - arg:
      - {{ i }}
    - tgt: "myproxy*'
    - parallel: True
{% endfor %}
```

I tried with parallel: True and False and both of them reported results.  I tried throwing an exception if `/tmp/dead` existed but that actually produced the stack trace on the salt master. The only two times I was able to get it only to report `ran on <one proxy>` was if a) I killed the salt-proxy service or b) in the function I did `exit()` which would shut the proxy minion down. It didn't matter that ping or alive would return False, orchestrate would still report `ran on myproxy-2, myproxy-1`.

I feel *very* confident in saying that you're observing is happening because either a) the service on either `nxs2-ork3` or `nxs1-ork1` is not running, or b) when it tries to run the commands it takes so long that the communication times out.

Basically - I don't think we're seeing a bug here, unless there's something horribly broken in the underlying napalm modules, but I think we'd see complete failures on both proxy minions instead. I'm going to go ahead and close this issue, since it doesn't appear to be a problem in Salt, but more of an environment issue. I'll see if we can figure out what's going on here, though! :+1:
